integrationTests/vm/txsFee/migrateDataTrie_test.go

```diff
@@ -0,0 +1,317 @@
+		// return after loading an extension node, not enough gas for the migration
+		gasLimit = uint64(50000)
+		migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
+		testGasConsumed(t, testContext, gasLimit, 40000)
+
+		// return after loading a branch node, not enough gas for the migration
+		gasLimit = uint64(70000)
+		migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
+		testGasConsumed(t, testContext, gasLimit, 60000)
+
+		// migrate 2 leaves, 5 loads + 2 stores = 200k gas
+		gasLimit = uint64(200000)
+		migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
+		testGasConsumed(t, testContext, gasLimit, 200000)
+
+		// migrate 2 leaves, 3 loads + 2 stores = 160k gas
+		gasLimit = uint64(200000)
+		migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
+		testGasConsumed(t, testContext, gasLimit, 160000)
+
+		// migrate 1 leaf, 2 loads + 1 store = 90k gas
+		gasLimit = uint64(200000)
+		migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
+		testGasConsumed(t, testContext, gasLimit, 90000)
+
+		// no leaf left to migrate, 1 load = 20k gas
+		gasLimit = uint64(200000)
+		migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
+		testGasConsumed(t, testContext, gasLimit, 20000)
+
+		err = dtr.Commit()
+		require.Nil(t, err)
+
+		rootHash, err = dtr.RootHash()
+		require.Nil(t, err)
+
+		dts, err = stats.GetTrieStats("", rootHash)
+		require.Nil(t, err)
+		require.Equal(t, uint64(0), dts.GetLeavesMigrationStats()[core.NotSpecified])
+		require.Equal(t, uint64(numDataTrieLeaves), dts.GetLeavesMigrationStats()[core.AutoBalanceEnabled])
+	})
+
+	t.Run("random trie - all leaves are migrated in multiple transactions", func(t *testing.T) {
+		t.Parallel()
+
+		testContext, err := vm.CreatePreparedTxProcessorWithVMsWithShardCoordinatorDBAndGas(enableEpochs, shardCoordinator, integrationTests.CreateMemUnit(), gasScheduleNotifier)
+		require.Nil(t, err)
+		defer testContext.Close()
+
+		sndNonce := uint64(0)
+		senderBalance := big.NewInt(math.MaxInt64)
+		_, _ = vm.CreateAccount(testContext.Accounts, sndAddr, sndNonce, senderBalance)
+
+		numLeaves := 10000
+		keyGenerator := func(i int) []byte {
+			return integrationTests.GenerateRandomSlice(32)
+		}
+		rootHash, allKeys := generateDataTrie(t, testContext, sndAddr, numLeaves, keyGenerator)
+		nonMigratedKeys := allKeys[1:]
+
+		acc := getAccount(t, testContext, sndAddr)
+		dtr, ok := acc.DataTrie().(common.Trie)
+		require.True(t, ok)
+		stats, ok := dtr.(statsCollector)
+		require.True(t, ok)
+
+		dts, err := stats.GetTrieStats("", rootHash)
+		require.Nil(t, err)
+
+		numNonMigratedLeaves := dts.GetLeavesMigrationStats()[core.NotSpecified]
+		numMigratedLeaves := dts.GetLeavesMigrationStats()[core.AutoBalanceEnabled]
+		require.Equal(t, uint64(numLeaves-1), numNonMigratedLeaves)
+		require.Equal(t, uint64(1), numMigratedLeaves)
+
+		gasLimit := uint64(500000000)
+		for numNonMigratedLeaves > 0 {
+			migrateDataTrie(t, testContext, sndAddr, gasPrice, gasLimit)
```

Comment:
can we have a counter with an absolute maximum value check? Something like: I know this will finish in 5-10 rounds. Ok, the test should not finish in 20 rounds as we might have a degradation.

Reply:
Done.